### DISCUSSION TITLE
How to include external jar libraries in OpenJML21

### DISCUSSION THREAD

**Discussion by jcrecio** [OpenJML/OpenJML#821] [General]
Hello,
I wonder given the current compilation system for OpenJML21, if I want to add an external dependency like for instance: **gson-2.8.1.jar** or **jackson-core-2.9.9.jar** how should I do it?

Thanks!

**Comment by davidcok** [comment]
A) If you are wanting to run either ESC or RAC on your own Java code along with a jar fi;e, you should be able to just include the jar file in the class path. Let me know the details if that does not work for you.

B) If you are wanting to extend OpenJML with new functionality, I don't have a quick recipe for you. The OpenJDK build system for the OpenJDK Java compiler is indeed complex -- it has a bootstrap build and then the "real" build, all made more complicated by the module system. But I will look into this and follow up.

**Comment by davidcok** [comment]
Which library and which version do you prefer?  e.g. jackson-core-2.9.9 is quite old.

You are going to have to be on Java 21 in any case.

- David

**Comment by jcrecio** [comment]
They idea is mainly, replicate the behaviour of this PR and after that upgrade all libraries to latest:
[Add command to serialize the typed AST](https://github.com/OpenJML/OpenJML/pull/691/commits)

Upgrade following libs to latest, is this anything feasible?

	<classpathentry kind="lib" path="otherlibs/gson-2.8.1.jar"/>
	<classpathentry exported="true" kind="lib" path="otherlibs/gson-2.8.5.jar"/>
	<classpathentry exported="true" kind="lib" path="otherlibs/jackson-annotations-2.9.9-javadoc.jar"/>
	<classpathentry exported="true" kind="lib" path="otherlibs/jackson-annotations-2.9.9.jar"/>
	<classpathentry exported="true" kind="lib" path="otherlibs/jackson-core-2.9.9-javadoc.jar"/>
	<classpathentry exported="true" kind="lib" path="otherlibs/jackson-core-2.9.9.jar"/>
	<classpathentry exported="true" kind="lib" path="otherlibs/jackson-databind-2.9.9-javadoc.jar"/>
	<classpathentry exported="true" kind="lib" path="otherlibs/jackson-databind-2.9.9.jar"/>
	<classpathentry kind="lib" path="otherlibs/reflections-0.9.12-SNAPSHOT.jar"/>
	<classpathentry exported="true" kind="lib" path="otherlibs/reflections-0.9.12-SNAPSHOT-jar-with-dependencies.jar"/>
	<classpathentry exported="true" kind="lib" path="otherlibs/jackson-jsog-1.1.1.jar"/>
	
	
Thanks David
